astar_utils/spectral_types.py

```diff
@@ -0,0 +1,150 @@
+    individual attribute, creation can only happen via the constructor string
+    described above.
+
+    One of the main features of this class (and indeed motivation for its
+    creation in the first place) is the ability to compare, and thus order,
+    instances of the class based on their two-component spectral (sub)class.
+    In this context, a "later" spectral type is considered "greater" than an
+    "earlier" one, i.e. O < B < A < F < G < K < M, which is consistent with the
+    convention of numerical subtypes, where 0 < 9 holds true. This is, however,
+    in contrast to the physical meaning of this parameter, which is correlated
+    with stellar effective temperature in reverse order, meaning T(A) > T(G)
+    and T(F0) > T(F9), by convention. On the other hand, in many visualisations
+    such as the Herzsprung-Russel diagram, it is common practice to represent
+    temperature on the x-axis in descending order, meaning a sorted list of
+    instances of this class will already have the typical order of such
+    diagrams.
+
+    In this context, the luminosity class (if any) is ignored for sorting and
+    comparison (<, >, <=, >=), as it represents a second physical dimension.
+    However, instances of this class may also be compared for equality (== and
+    !=), in which case all three attributes are considered.
+
+    Attributes
+    ----------
+    spectral_class : str
+        Main spectral class (OBAFGKM).
+    spectral_subclass : str or None
+        Numerical spectral subclass (0.0-9.9).
+    luminosity_class : str or None
+        Roman numeral luminosity class (I-V).
+
+    Examples
+    --------
+    >>> from astar_utils import SpectralType
+    >>> spt = SpectralType("A0V")
+    >>> spt.spectral_class
+    'A'
+
+    >>> spt.spectral_subclass
+    0.0
+
+    >>> spt.luminosity_class
+    'V'
+
+    >>> spts = [SpectralType(s) for s in
+    ...         ["G2", "M4.0", "B3", "F8", "K6.5",
+    ...          "A", "A0", "A9", "O8"]]
+    >>> sorted(spts)  # doctest: +NORMALIZE_WHITESPACE
+    [SpectralType('O8'),
+     SpectralType('B3'),
+     SpectralType('A0'),
+     SpectralType('A'),
+     SpectralType('A9'),
+     SpectralType('F8'),
+     SpectralType('G2'),
+     SpectralType('K6.5'),
+     SpectralType('M4')]
+
+    Note that a missing spectral subtype is considered as 5 (middle of the
+    main spectral class) in the context of sorting, as shown in the example
+    with the spectral type "A" ending up between "A0" and "A9".
+    """
+
+    spectral_class: str = field(init=False, default="")
+    spectral_subclass: float | None = field(init=False, default=None)
+    luminosity_class: str | None = field(init=False, default=None)
+    spectype: InitVar[str]
+    _cls_order: ClassVar = "OBAFGKM"  # descending Teff
+    _regex: ClassVar = re.compile(
+        r"^(?P<spec_cls>[OBAFGKM])(?P<sub_cls>\d(?:\.\d)?)?"
+        "(?P<lum_cls>I{1,3}|IV|V)?$", re.A | re.I)
```

Comment:
Maybe, just to be ultra extra diligent:
```suggestion
        "(?P<lum_cls>I{1,3}|IV|V)?\s*$", re.A | re.I)
```
Because, who knows, maybe someone adds some extra whitespace at the end. :wink:
J/k, probably don't do this, as it does make the regex a bit more complex to handle other peoples bugs.

Reply:
I see where you're going with that 😂